tests/python/test_py_dataclass.py

```diff
@@ -864,3 +864,50 @@ def k1(p1: P1) -> None:
     assert d[0] == 55
     assert e[0] == 66
     assert f[0] == 77
+
+
+@test_utils.test()
+def test_template_mapper_cache_no_collision():
+    @dataclass(frozen=True)
+    class MyStruct:
+        value: ti.types.ndarray()
+
+    @ti.kernel
+    def my_kernel(my_struct_1d: MyStruct, my_struct_2d: MyStruct) -> None:
+        for i in ti.ndrange(my_struct_1d.value.shape[0]):
+            my_struct_1d.value[i] += 1
+        for i, j in ti.ndrange(my_struct_2d.value.shape[0], my_struct_2d.value.shape[1]):
+            my_struct_2d.value[i, j] += 1
+
+    value = ti.ndarray(ti.i32, shape=(1,))
+    value.fill(0)
+    my_struct_1d = MyStruct(value=value)
+    value = ti.ndarray(ti.f32, shape=(1, 2))
+    value.fill(0.0)
+    my_struct_2d = MyStruct(value=value)
+
+    my_kernel(my_struct_1d, my_struct_2d)
+    assert my_struct_1d.value[0] == 1
+    assert my_struct_2d.value[0, 0] == 1.0
+    assert my_struct_2d.value[0, 1] == 1.0
+
+
+@test_utils.test()
+def test_template_mapper_cache_ignore_slots():
+    @dataclass(frozen=True, slots=True)
+    class MyStruct:
+        value: ti.types.ndarray()
+
+    @ti.kernel
+    def my_kernel(my_struct: MyStruct) -> None:
+        for i in ti.ndrange(my_struct.value.shape[0]):
+            my_struct.value[i] += 1
+
+    value = ti.ndarray(ti.i32, shape=(1,))
+    value.fill(0)
+    my_struct = MyStruct(value=value)
+
+    my_kernel(my_struct)
+    assert my_struct.value[0] == 1
+    my_kernel(my_struct)
```

Comment:
Don't we need to test what happens if we create a brand new instance of MyStruct, as we did in the first test? What will happen? Should that throw an exception?

Reply:
(I think I need to add Sonarqube, to check coverage, at some point :) )

Reply:
Because it is checking something completely different. I'm checking that dataclasses that enable slots are ignored by "caching" mechanism. Without caching, we are back to the "usual" dataclass, which is already thoroughly tested.

Reply:
I unified tests for slots=True and False.